src/core/thread/network_data_leader.cpp

```diff
@@ -67,6 +67,117 @@ void LeaderBase::Reset(void)
+                                 bool           aServerStable,
+                                 uint8_t &      aSID)
+{
+    otError error = OT_ERROR_NOT_FOUND;
+
+    Iterator        iterator = kIteratorInit;
+    otServiceConfig serviceConfig;
+
+    while (GetNextService(iterator, serviceConfig) == OT_ERROR_NONE)
+    {
+        if (aEnterpriseNumber == serviceConfig.mEnterpriseNumber &&
+            aServiceDataLength == serviceConfig.mServiceDataLength &&
+            memcmp(aServiceData, serviceConfig.mServiceData, aServiceDataLength) == 0 &&
+            aServerStable == serviceConfig.mServerConfig.mStable)
+        {
+            aSID = serviceConfig.mServiceID;
+            ExitNow(error = OT_ERROR_NONE);
+        }
+    }
+
+exit:
+    return error;
+}
+
+#if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
+otError LeaderBase::GetBackboneRouterPrimary(BackboneRouter::BackboneRouterConfig &aConfig,
+                                             BackboneRouter::Leader::State &       aState)
+{
+    otError                         error          = OT_ERROR_NOT_FOUND;
+    uint8_t                         serviceData    = ServiceTlv::kServiceDataBackboneRouter;
+    const ServerTlv *               rvalServerTlv  = NULL;
+    const BackboneRouterServerData *rvalServerData = NULL;
+    ServiceTlv *                    serviceTlv;
+    NetworkDataTlv *                subCur;
+    NetworkDataTlv *                subEnd;
+
+    aState = BackboneRouter::Leader::kStateNone;
+
+    serviceTlv = Get<Leader>().FindService(THREAD_ENTERPRISE_NUMBER, &serviceData, sizeof(serviceData));
+
+    if (serviceTlv == NULL)
+    {
+        if (aConfig.mServer16 != Mac::kShortAddrInvalid)
+        {
+            aConfig.mServer16 = Mac::kShortAddrInvalid;
+            aState            = BackboneRouter::Leader::kStateRemoved;
+        }
+
+        ExitNow();
+    }
+
+    subCur = serviceTlv->GetSubTlvs();
+    subEnd = serviceTlv->GetNext();
+
+    while (subCur < subEnd)
+    {
+        ServerTlv *                     serverTlv;
+        const BackboneRouterServerData *serverData;
+
+        VerifyOrExit((subCur + 1) <= subEnd && subCur->GetNext() <= subEnd, error = OT_ERROR_PARSE);
+
+        serverTlv  = static_cast<ServerTlv *>(subCur);
+        serverData = reinterpret_cast<const BackboneRouterServerData *>(serverTlv->GetServerData());
+
+        if (rvalServerTlv == NULL ||
+            (serverTlv->GetServer16() == Mle::Mle::Rloc16FromRouterId(Get<Mle::MleRouter>().GetLeaderId())) ||
+            serverData->GetSequenceNumber() > rvalServerData->GetSequenceNumber() ||
+            serverTlv->GetServer16() > rvalServerTlv->GetServer16())
+        {
+            rvalServerTlv  = const_cast<ServerTlv *>(serverTlv);
+            rvalServerData = serverData;
+        }
+
+        subCur = subCur->GetNext();
+    }
+
+    error = OT_ERROR_NONE;
+
+    if (aConfig.mServer16 == Mac::kShortAddrInvalid)
```

Comment:
How about keeping this method to simply retrieving the primary bbr info and moving the state change logic to `Leader::Update()`?

Reply:
seems more reasonable in code structure. will update.